File: ch05-eval-harness/R/run_eval.R
#!/usr/bin/env Rscript
# Minimal eval harness (offline mock scorer).
# For live LLM scoring, replace mock_predict() with your API wrapper.

args <- commandArgs(trailingOnly = TRUE)
golden_path <- if (length(args) >= 1) args[[1]] else "golden/dev_20.csv"
prompt_path <- if (length(args) >= 2) args[[2]] else "prompts/classify_v1.md"

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
})

golden <- read_csv(golden_path, show_col_types = FALSE)
prompt <- paste(readLines(prompt_path, warn = FALSE), collapse = "\n")

# Deterministic mock: keyword rules (NOT a real model). For teaching the harness loop.
mock_predict <- function(text) {
  t <- text
  if (grepl("선포|동의|표결|속기록|의사일정|정회", t)) return("절차")
  if (grepl("규탄|위선|무능|왜곡|비판|공격|발목", t)) return("정쟁")
  "정책"
}

pred <- golden %>%
  mutate(
    model_label = vapply(text, mock_predict, character(1)),
    correct = model_label == human_label
  )

acc <- mean(pred$correct)
cat(sprintf("N=%d  accuracy=%.3f  (mock keyword model)\n", nrow(pred), acc))
cat("Confusion (human x model):\n")
print(table(human = pred$human_label, model = pred$model_label))

dir.create("output", showWarnings = FALSE)
out <- file.path("output", "eval_dev20_mock.csv")
write_csv(pred, out)
cat("wrote", out, "\n")
cat("Prompt fingerprint (first 60 chars):", substr(prompt, 1, 60), "...\n")

File: ch05-eval-harness/python/run_eval.py
#!/usr/bin/env python3
"""Minimal eval harness with offline mock scorer."""
from __future__ import annotations

import argparse
import csv
import re
from collections import Counter
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


def mock_predict(text: str) -> str:
    if re.search(r"선포|동의|표결|속기록|의사일정|정회", text):
        return "절차"
    if re.search(r"규탄|위선|무능|왜곡|비판|공격|발목", text):
        return "정쟁"
    return "정책"


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--golden", type=Path, default=ROOT / "golden" / "dev_20.csv")
    ap.add_argument("--prompt", type=Path, default=ROOT / "prompts" / "classify_v1.md")
    args = ap.parse_args()
    prompt = args.prompt.read_text(encoding="utf-8")
    rows = list(csv.DictReader(args.golden.open(encoding="utf-8")))
    out_rows = []
    ok = 0
    conf = Counter()
    for r in rows:
        pred = mock_predict(r["text"])
        correct = pred == r["human_label"]
        ok += int(correct)
        conf[(r["human_label"], pred)] += 1
        out_rows.append({**r, "model_label": pred, "correct": correct})
    n = len(rows)
    print(f"N={n}  accuracy={ok/n:.3f}  (mock keyword model)")
    print("confusion human->model:", dict(conf))
    out = ROOT / "output" / "eval_dev20_mock.csv"
    out.parent.mkdir(exist_ok=True)
    with out.open("w", encoding="utf-8", newline="") as f:
        w = csv.DictWriter(f, fieldnames=list(out_rows[0].keys()))
        w.writeheader()
        w.writerows(out_rows)
    print("wrote", out)
    print("prompt head:", prompt[:60].replace("\n", " "), "...")


if __name__ == "__main__":
    main()
